pyrk/th_system.py

```diff
@@ -104,3 +107,262 @@ def custom(self, t_b, t_env, res):
+                to_ret -= Qadv/cap/component.vol.magnitude
+                assert Qadv >= 0, '''at step %d, %s is heating the system
+                by %f watts??? Tin is %f, tout is %f, tcomp is %f''' % (
+                    t_idx, component.name, Qadv, d['t_in'].magnitude,
+                    (component.T[t_idx]*2 - d['t_in']).magnitude,
+                    component.T[t_idx].magnitude)
+            return to_ret*units.kelvin/units.seconds
+
+    def BC_center(self, component, t_idx):
+        '''Qconduction from the center of a sphere to the first boundary
+        (conduction without interface) in watts/meter**3
+        :param component: name of the inner most component(mesh element)
+        :type component: str
+        :param t_idx: timestep at which Qconduction is calculated
+        :type t_idx: int
+        :return : Qcondction
+        :rtype:float, dimensionless
+        '''
+        conductivity = component.k.magnitude
+        T_b = component.T[t_idx].magnitude
+        dr = (component.ro - component.ri).magnitude
+        return (conductivity*T_b)/(dr**2)
+
+    def convBoundary(self, component, t_b, t_env, h, R):
+        '''calculate heat transfer through convective boundray condition
+        :param component: name of the outer most solid component
+        :type component: str
+        :param t_b: temperature of the body
+        :type t_b: float
+        :param t_env: temperature of the environment(fluid)
+        :type t_env: float
+        :param h: convective heat transfer coefficient
+        :type h: float
+        :param R: outer radius of the component
+        :type R: float
+        :return: dimensionless quantity of Qconv
+        :rtype: float
+        '''
+        r_b = component.ro.magnitude
+        k = component.k.magnitude
+        dr = component.ri.magnitude-component.ro.magnitude
+        T_R = (-h.magnitude/k*t_env + t_b/dr)/(1/dr-h.magnitude/k)
+        to_ret = 1/r_b*k*(r_b*t_b-R.magnitude*T_R)/dr**2
+        return to_ret
+
+    def heatgenFVM(self, component, power, omegas):
+        return (power*component.power_tot.magnitude *
+                (1-self.kappa) + sum(omegas))/component.vol.magnitude
+
+    def conductionFVM(self, component, env, t_idx):
+        """
+        heat transfer by conduction(watts/m3)
+        :param component: name of the component
+        :type component: str
+        :param env: name of the environment
+        :type env: str
+        :param t_idx: time step that conduction heat is computed
+        :type t_idx: int
+        :return: Qond, dimemsionless quantity
+        :rtype: float
+        """
+        T_b = component.T[t_idx].magnitude
+        T_env = env.T[t_idx].magnitude
+        r_b = component.ro.magnitude
+        r_env = env.ro.magnitude
+        dr = (component.ro-component.ri).magnitude
+        k = component.k.magnitude
+        return k/r_b * (r_b * T_b - r_env * T_env)/(dr**2)
+
+    def advection(self, t_out, t_in, m_flow, cp):
```

Comment:
In the rest of the code, we have adopted a (somewhat unpleasant, but fairly clear) convention that little t means time and big T means temperature. This doesn't match that. I'm not sure if that matters to you. Just something to point out.